src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs

```diff
@@ -0,0 +1,465 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    ///  Represents a JSON object.
+    /// </summary>
+    public sealed class JsonObject : JsonNode, IEnumerable<KeyValuePair<string, JsonNode>>
+    {
+        private readonly Dictionary<string, JsonNode> _dictionary;
+        private readonly DuplicatePropertyNameHandling _duplicatePropertyNameHandling;
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class representing the empty object.
+        /// </summary>
+        /// <param name="duplicatePropertyNameHandling">Specifies the way of handling duplicate property names.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided manner of handling duplicates does not exist.
+        /// </exception>
+        public JsonObject(DuplicatePropertyNameHandling duplicatePropertyNameHandling = default)
+        {
+            if(!Enum.IsDefined(typeof(DuplicatePropertyNameHandling), duplicatePropertyNameHandling))
+            {
+                throw new ArgumentException(SR.InvalidDuplicatePropertyNameHandling);
+            }
+
+            _dictionary = new Dictionary<string, JsonNode>();
+            _duplicatePropertyNameHandling = duplicatePropertyNameHandling;
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class representing provided set of JSON properties.
+        /// </summary>
+        /// <param name="jsonProperties">>Properties to represent as a JSON object.</param>
+        /// <param name="duplicatePropertyNameHandling">Specifies the way of handling duplicate property names.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided collection contains duplicates if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public JsonObject(IEnumerable<KeyValuePair<string, JsonNode>> jsonProperties, DuplicatePropertyNameHandling duplicatePropertyNameHandling = default)
```

Comment:
While `default` and `DuplicatePropertyNameHandling.Replace` are logically the same, the latter helps with clarity (and avoids needing to look at the second type to remember which thing was zero).

Also, these lines are pretty long (longer than the default github review width); it'd be nice to people who don't want to use browser stylesheet overrides to chop these lines down.

```suggestion
        public JsonObject(
            IEnumerable<KeyValuePair<string, JsonNode>> jsonProperties,
            DuplicatePropertyNameHandling duplicatePropertyNameHandling = DuplicatePropertyNameHandling.Replace)
```